transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollSocketHalfClosedReproTest.java

```diff
@@ -0,0 +1,188 @@
+                @Override
+                protected void initChannel(Channel ch) throws Exception {
+                    logger.info("{} Client Channel", ch);
+                    ch.config().setAutoClose(false);
+                    ch.pipeline().addLast(new ChannelInboundHandlerAdapter() {
+                        private int bytesRead;
+                        private int bytesSinceReadComplete;
+
+                        @Override
+                        public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                            ByteBuf buf = (ByteBuf) msg;
+                            bytesRead += buf.readableBytes();
+                            bytesSinceReadComplete += buf.readableBytes();
+                            buf.release();
+                        }
+
+                        @Override
+                        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+                            if (evt == ChannelInputShutdownEvent.INSTANCE) {
+                                clientHalfClosedLatch.countDown();
+                            } else if (evt == ChannelInputShutdownReadComplete.INSTANCE) {
+                                clientHalfClosedAllBytesRead.countDown();
+                                ctx.close();
+                            }
+                        }
+
+                        @Override
+                        public void channelReadComplete(ChannelHandlerContext ctx) {
+                            if (bytesSinceReadComplete == 0) {
+                                clientZeroDataReadCompletes.incrementAndGet();
+                            } else {
+                                bytesSinceReadComplete = 0;
+                            }
+                            clientReadCompletes.incrementAndGet();
+                            if (bytesRead == totalServerBytesWritten) {
+                                // Bounce this through the event loop to make sure it happens after we're done
+                                // with the read operation.
+                                ch.eventLoop().execute(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        clientReadAllDataLatch.countDown();
+                                    }
+                                });
+                            } else {
+                                ch.read();
+                            }
+                        }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+                            ctx.close();
+                        }
+                    });
+                    ch.read();
+                }
+            });
+
+            serverChannel = sb.bind().sync().channel();
+            clientChannel.set(cb.connect(serverChannel.localAddress()).sync().channel());
```

Comment:
Why we need this AtomicReference at all ? Seems like We could just reference the Channel directly in this scope

Reply:
We don't at this point. I'll remove it before it's merged, presuming thats what we decide to do.